File: src/main/python/sortByLastname.py
#=====================================================================                
# main 
#=====================================================================   
names = ["Jonathan Smith", "Bill O'Reilly", "Thomas Sowell Sr.", "Jonathan Smith", "Rose Friedmann"]
nameSet = set(names)

print ", ".join(str(name) for name in nameSet)

nameTuples = []
for name in nameSet:
    words = name.split()
    nameTuples = nameTuples + [(name, words[1])] 

print nameTuples

# sort by lastName
sortedNameTuples = sorted(nameTuples, key=lambda name: name[1])

print "-----"
print ", ".join(name[0] for name in sortedNameTuples)

File: src/main/java/org/zfun/leetcode/SortByLastName.java
package org.zfun.leetcode;

import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class SortByLastName {

    // Print out the name sorted by lastName, which is the second word without duplicates
    public static void main(String[] args) {
        List<String> names = new LinkedList<>(Arrays.asList("Jonathan Smith", "Bill O'Reilly", "Thomas Sowell Sr.",
                "Jonathan Smith", "Rose Friedmann"));
        Set<String> nameSet = new HashSet<>(names);
        // sort
        List<String> sortedNames = nameSet.stream()
                .sorted((n1, n2) -> {
                    String[] tokens1 = n1.split(" ");
                    String[] tokens2 = n2.split(" ");
                    return tokens1[1].compareTo(tokens2[1]);
                })
                .collect(Collectors.toList());
        // print
        System.out.println(sortedNames.stream()
            .map(Object::toString)
            .collect(Collectors.joining(",","[","]"))
            )
        ;
        
    }
}
